### ISSUE TITLE
Missing attributes on tkinter.Events, tkinter.ttk.Frame, and tkinter.ttk.Treeview

### ISSUE THREAD

**Issue by jcrmatos** [python/typeshed#1631]
Hello,

Mypy complains that tkinter.Event doesn't have x_root/y_root attributes, which is incorrect.
Here is an example (abridged) of code that mypy complains about.

```
...
    def show(self, event):
        # Position the tooltip below and to the rigth of the cursor (16x16).
        # Mypy returns an error Event has no attr x_root/y_root.
        x_pos = event.x_root + 16  # type: ignore
        y_pos = event.y_root + 16  # type: ignore
        self.geometry('+{0}+{1}'.format(x_pos, y_pos))

        self.deiconify()
        self.visible = True

        self.after(self.delay, self.hide)  # Milliseconds.
...
widget.bind('<Enter>', self.show)
...
```

Best regards,

JM

**Comment by jcrmatos** [comment]
Hello,

Forgot to say that it also doesn't recognize the x, y, and widget attributes.
I'm using Python 3.5.3 and mypy 0.521.

Best regards,

JM

**Comment by jcrmatos** [comment]
Hello,

Just checked that it also complains about tkinter.ttk.Frame.master and tkinter.ttk.Treeview.master attributes.

Best regards,

JM

**Comment by JelleZijlstra** [comment]
Feel free to open a PR adding those! I don't think the stubs for tkinter have had a lot of attention.

**Comment by Akuli** [comment]
Tkinter stubs use a lot of `Any`, and they generate lots of mypy errors with `mypy --strict`.

I have written [my own tkinter stubs](https://github.com/Akuli/porcupine/tree/a6925e617/my_stubs), but I don't think that creating a pull request that adds them to typeshed is a good idea. I explained why in the README of my stubs, but the tl;dr is:
- My stubs are missing many things, and completing them perfectly would be a lot more work than I'm interested in doing. I'd rather add support for features as they are actually needed.
- My stubs contain some copy/pasta due to how tkinter's widget options work, and the copy/pasta amount would be pretty bad if my stubs supported every feature of tkinter.
- Putting my stubs to typeshed would break tkinter code that currently typechecks correctly in mypy.
- My stubs use Python 3.8 features. I guess typeshed must be more backwards compatible than that?
- My stubs would be more work to maintain than what's currently in typeshed. New features would need to be added every time a new release of Tk comes out, even when the duck-typed tkinter code doesn't need changes.

**Comment by srittau** [comment]
We are grateful for any improvements that can be made to our stubs. Stubs generally follow Python 3.7 syntax. For positional only arguments, we need to use the `__x` syntax, instead of being able to use the forward slash.

**Comment by Akuli** [comment]
I wonder how many people are actually using mypy and tkinter together, i.e. how bad it would be to break backwards compatibility in tkinter stubs. At least I'm quite confident that nobody is trying to use `mypy --strict` and tkinter together, because that just doesn't work.

**Comment by Akuli** [comment]
I guess I should clarify that "breaking backwards compatibility" means need to change type hints and add asserts to make mypy checks pass again. On the other hand, the newly generated mypy errors could also help with discovering bugs.

**Comment by hauntsaninja** [comment]
tkinter stubs are the most poorly supported stdlib stubs in typeshed, like Sebastian I'd be grateful for improvements to typeshed.

Just a scattered reply to some of your concerns raised here and at https://github.com/Akuli/porcupine/tree/a6925e617/my_stubs#why-not-pull-request

> My stubs use typing.Literal generously. That's new in Python 3.8, but typeshed stubs need to be backwards compatible.

typeshed stubs are able to use Literal (and it works for Python 3.7 and earlier as well). You just have to import it from `typing_extensions` (all type checkers understand this pattern).

> Any code using constants like tkinter.END would need to be changed to use strings like 'end' instead or I would need to add support for tkinter.END and similar constants.

> If a method is missing from my stubs but present in what's currently in typeshed, then all mypy-checked code using that method will stop working.

This is the kind of backward incompatibility I'd very much prefer not to have. But it's totally okay to introduce new type errors in user's code from type narrowing that result in true positives (or useful warnings) a lot of the time.

However, typeshed does err on the side of avoiding false positives, so for instance, we'd prefer to have a loosely typed `__getitem__` over omitting it.

> If someone calls a tkinter method in a way that I have never used, such as after(ms) in the above example, then they would start getting mypy errors.

In my opinion, this kind of issue is preferred to not having any definition or having a mostly untyped definition.

> In the past, typeshed contributors decided to not support '??' in tkinter.Event (see above). My tkinter.Event does that "right" and supports '??', and a lot of assert thing != '??' would need to be added to mypy checked code to make it work with my stubs.

As the reviewer of that PR, I still stand by what I said then: "Re "??", I'll trust [the contributor] on whether `Union[str, whatever]` is a more helpful type than `whatever`. Note that generally typeshed prefers to stick to implementation, so if this causes issues down the road, we'll change it."

Generally speaking, tkinter is a tricky case for me, since I don't have experience with using the library and it's not a type-friendly design. This means we're likely to have imperfect types no matter what we do and it's hard for me to ascertain what the right tradeoff is.
I think what to do here might come down to the specifics of what's missing from your stubs. We'd certainly accept PRs that improved the existing stubs based on yours, but it sounds like you might have in mind wholesale swapping out of the existing stubs for your stubs, in which case we might want to be a little more circumspect and/or do the work of merging the two.

**Comment by Akuli** [comment]
>>Any code using constants like tkinter.END would need to be changed to use strings like 'end' instead or I would need to add support for tkinter.END and similar constants.
>>...
>
>This is the kind of backward incompatibility I'd very much prefer not to have.

I will add type hints for `tkinter.END` and similar constants if I actually create a pull request. Some people like them, and this comes down to opinions. I don't want to shove my opinions into other people's code.

> As the reviewer of that PR, I still stand by what I said then: "Re "??", I'll trust [the contributor] on whether Union[str, whatever] is a more helpful type than whatever. 

I would dislike `Union[str, whatever]` myself too, but my code does `Union[str, Literal['??']]`. I think that's reasonable, and actually checking for that leads to a readable `assert event.thing != '??'` as opposed to a confusing `assert not isinstance(event.thing, str)`. I guess `Literal['??']` wasn't considered in the tkinter Event field pull request.

>Generally speaking, tkinter is a tricky case for me, since it's not a type-friendly design... This means we're likely to have imperfect types no matter what we do...

Tkinter is stringly typed because it's a wrapper for a stringly typed programming language. I really like using `Literal` with tkinter.

>...and I don't have experience with using the library.

I'm more than familiar with tkinter.

>I think what to do here might come down to the specifics of what's missing from your stubs.

If I'll actually make a pull request, then I won't delete any existing `Any`-typed methods without adding in "correctly" typed methods that replace those. I might leave a bunch of the `Any`-typed methods under a `TODO` comment or something, and similarly add `TODO` comments for missing option names.

**Comment by Akuli** [comment]
Copy/pasta will be another problem. How much copy/pasta is acceptable in stub files? I looked at `subprocess.Popen` stubs yesterday and they have several lines copy/pasted and slightly modified 4 times.

With the current mypy features that I'm aware of, every tkinter widget option would have quite a lot of copy/pasta:
- `__init__` kwarg
- `configure` kwarg
- `config` kwarg (`config` is an alias for `configure`)
- `cget` argument of type `Literal['option name']`
- `__setitem__` argument
- `__getitem__` argument (`__getitem__` is an alias for `cget`)
- Copy/pasting all these to inherited classes

Much of the copy/pasta could be avoided with nicer mypy features that are currently not implemented:
- `__init__`, `configure`, `config` kwargs: https://github.com/python/mypy/issues/4441
- Inheritance issues: (https://github.com/python/mypy/issues/9133)
- Typing aliased methods: some way to specify "these two methods have exactly the same type"

If **all** this was available, then there wouldn't be almost no copy/pasta left for each option:
- `TypedDict` or something else for kwargs of `__init__`, `configure`, `config`
- `__setitem__` with exactly the same types as `__init__`, `configure`, `config` (unless there's some way to borrow the `__setitem__` of a `TypedDict`)
- `cget` and `__setitem__` (the types of the options are different for these than for `__init__`, `config`, `configure`

Meanwhile, is it acceptable to write a script that generates the copy/pasta for stub files?

This affects only two stub files, `tkinter/__init__.pyi` and `tkinter/ttk.pyi`.

**Comment by Akuli** [comment]
>As the reviewer of that PR, I still stand by what I said then: "Re "??", I'll trust [the contributor] on whether Union[str, whatever] is a more helpful type than whatever.

I'm starting to agree, because event fields being `'??'` just doesn't come up much in practice. If you write an event handler for a keyboard event, then it's very unlikely that you try to do anything with `event.width`. On the other hand, if you write an event handler for a window resizing event, then there's no way how `event.width` could be `'??'` and there's no actual reason to check that it's not `'??'`. That is, `'??'` is not actually a problem, and there's no need to "solve" it with `Literal['??']` or `str`. Practicality beats purity here.

**Comment by hauntsaninja** [comment]
Generally speaking typeshed is quite accommodating of copy-pasta. Although mypy is the most visible type checker, typeshed is the lowest common denominator for a lot of different tools, so sometimes dumb (potentially generated) copy pasta is the best way forward to bridge the various gaps. If it's understandable copy pasta and less than a couple thousand lines long I wouldn't worry too much.

> My stubs allow for example label['text'], which comes out as a string, but not label['blablabla'] or label[variable_with_type_str] ... What would be the preferable thing for this? ... my "large" tkinter project ... didn't have any issues?

typeshed pretty strongly prefers not to have false positives, so my instinct would be to use Literal overloads and then have str return Any. But again, I don't know what's idiomatic for tkinter... If you feel it would catch substantially more typos and false positives are very rare, personally I'd be willing to accept it for now, and only change it if people complain.

If it would make things much less awful, you might be able to get away with aliasing methods to each other to replicate their types. (pytype and other tools sometimes doesn't like this and complain)

**Comment by Akuli** [comment]
Look at the `tkinter.font` stub from #4350 to get an idea of what the copy/pasta is like. The difference is that the copy/pasta for widget options is similar except that there's about 2000 lines of it.

Currently I have a script that generates the widget option copy/pasta. The auto-generated copy/pasta looks like this:

```
class Toplevel(BaseWidget, Wm):
    if sys.version_info >= (0,):  # AUTOGENERATED scripts/tkinter_options_script.py tkinter.Toplevel
        # fmt: off
        def __init__(self, master: Optional[Misc] = ..., cnf: Dict[str, Any] = ..., *, background: str = ..., bd: _ScreenUnits = ..., bg: str = ..., borderwidth: _ScreenUnits = ..., class_: str = ..., colormap: Union[Literal['new', ''], Misc] = ..., container: bool = ..., cursor: Union[str, Tuple[str], Tuple[str, str], Tuple[str, 
str, str], Tuple[str, str, str, str]] = ..., height: _ScreenUnits = ..., highlightbackground: str = ..., highlightcolor: str = ..., highlightthickness: _ScreenUnits = ..., menu: Menu = ..., padx: _ScreenUnits = ..., pady: _ScreenUnits = ..., relief: Literal["raised", "sunken", "flat", "ridge", "solid", "groove"] = ..., takefocus: 
Union[bool, Literal[''], Callable[[str], Optional[bool]]] = ..., use: int = ..., visual: Union[str, Tuple[str, int]] = ..., width: _ScreenUnits = ...) -> None: ...
        @overload
        def configure(self, cnf: Dict[str, Any] = ..., *, background: str = ..., bd: _ScreenUnits = ..., bg: str = ..., borderwidth: _ScreenUnits = ..., class_: str = ..., colormap: Union[Literal['new', ''], Misc] = ..., container: bool = ..., cursor: Union[str, Tuple[str], Tuple[str, str], Tuple[str, str, str], Tuple[str, str, st
r, str]] = ..., height: _ScreenUnits = ..., highlightbackground: str = ..., highlightcolor: str = ..., highlightthickness: _ScreenUnits = ..., menu: Menu = ..., padx: _ScreenUnits = ..., pady: _ScreenUnits = ..., relief: Literal["raised", "sunken", "flat", "ridge", "solid", "groove"] = ..., takefocus: Union[bool, Literal[''], Call
able[[str], Optional[bool]]] = ..., use: int = ..., visual: Union[str, Tuple[str, int]] = ..., width: _ScreenUnits = ...) -> None: ...
        @overload
        def configure(self) -> Dict[str, Tuple[str, str, str, Any, Any]]: ...
        @overload
        def configure(self, cnf: Literal['background', 'bd', 'bg', 'borderwidth', 'class', 'colormap', 'container', 'cursor', 'height', 'highlightbackground', 'highlightcolor', 'highlightthickness', 'menu', 'padx', 'pady', 'relief', 'takefocus', 'use', 'visual', 'width']) -> Tuple[str, str, str, Any, Any]: ...
        def cget(self, key: Literal['background', 'bd', 'bg', 'borderwidth', 'class', 'colormap', 'container', 'cursor', 'height', 'highlightbackground', 'highlightcolor', 'highlightthickness', 'menu', 'padx', 'pady', 'relief', 'takefocus', 'use', 'visual', 'width']) -> Any: ...
        @overload
        def __setitem__(self, key: Literal['background'], val: str) -> None: ...
        @overload
        def __setitem__(self, key: Literal['bd'], val: _ScreenUnits) -> None: ...
        @overload
        def __setitem__(self, key: Literal['bg'], val: str) -> None: ...
        @overload
        def __setitem__(self, key: Literal['borderwidth'], val: _ScreenUnits) -> None: ...
        @overload
        def __setitem__(self, key: Literal['class'], val: str) -> None: ...
        @overload
        def __setitem__(self, key: Literal['colormap'], val: Union[Literal['new', ''], Misc]) -> None: ...
        @overload
        def __setitem__(self, key: Literal['container'], val: bool) -> None: ...
        @overload
        def __setitem__(self, key: Literal['cursor'], val: Union[str, Tuple[str], Tuple[str, str], Tuple[str, str, str], Tuple[str, str, str, str]]) -> None: ...
        @overload
        def __setitem__(self, key: Literal['height'], val: _ScreenUnits) -> None: ...
        @overload
        def __setitem__(self, key: Literal['highlightbackground'], val: str) -> None: ...
        @overload
        def __setitem__(self, key: Literal['highlightcolor'], val: str) -> None: ...
        @overload
        def __setitem__(self, key: Literal['highlightthickness'], val: _ScreenUnits) -> None: ...
        @overload
        def __setitem__(self, key: Literal['menu'], val: Menu) -> None: ...
        @overload
        def __setitem__(self, key: Literal['padx'], val: _ScreenUnits) -> None: ...
        @overload
        def __setitem__(self, key: Literal['pady'], val: _ScreenUnits) -> None: ...
        @overload
        def __setitem__(self, key: Literal['relief'], val: Literal["raised", "sunken", "flat", "ridge", "solid", "groove"]) -> None: ...
        @overload
        def __setitem__(self, key: Literal['takefocus'], val: Union[bool, Literal[''], Callable[[str], Optional[bool]]]) -> None: ...
        @overload
        def __setitem__(self, key: Literal['use'], val: int) -> None: ...
        @overload
        def __setitem__(self, key: Literal['visual'], val: Union[str, Tuple[str, int]]) -> None: ...
        @overload
        def __setitem__(self, key: Literal['width'], val: _ScreenUnits) -> None: ...
        config = configure
        __getitem__ = cget
        # fmt: on
```

Here `if sys.version_info >= (0,)` is a code-folding-editor and linter friendly way to mark where the copy/pasta begins and ends (flake8 doesn't like `if True:` in stub files).

Is this acceptable for creating a pull request to typeshed?

**Comment by Akuli** [comment]
Nevermind, there will actually be a lot less copy/pasta.

**Comment by Akuli** [comment]
This issue can be closed:
- Event object attributes were added in #4200
- missing `master` attributes were fixed in #4468 
- my stubs have mostly been pull requested into typeshed

**Comment by srittau** [comment]
Thanks for all your work, @Akuli!
